src/verify/verify.js

```diff
@@ -0,0 +1,167 @@
+        fileName: fileName,
+        contractName: contractName
+    }
+}
+
+let recompile = async function(metadata, sources) {
+    let reformatted = reformatMetadata(metadata, sources)
+    let input = reformatted.input
+    let fileName = reformatted.fileName
+    let contractName = reformatted.contractName
+
+    console.log('Re-compiling ' + fileName + ':' + contractName + ' with Solidity ' + metadata['compiler']['version'])
+    console.log('Retrieving compiler...')
+    let solcjs = await new Promise((resolve, reject) => {
+        solc.loadRemoteVersion('v' + metadata['compiler']['version'], (error, soljson) => {
+            if (error) {
+                reject()
+            } else {
+                resolve(soljson)
+            }
+        })
+    })
+    console.log('Compiling...');
+    let output = JSON.parse(solcjs.compile(JSON.stringify(input)));
+    return {
+        bytecode: output['contracts'][fileName][contractName]['evm']['bytecode']['object'],
+        deployedBytecode: '0x' + output['contracts'][fileName][contractName]['evm']['deployedBytecode']['object'],
+        metadata: output['contracts'][fileName][contractName]['metadata'].trim()
+    }
+}
+
+let web3 = new Web3('https://ropsten.infura.io/v3/891fe57328084fcca24912b662ad101f')
+
+let run = async function(addr, payload)
+{
+    let data = await getBytecodeMetadataAndSources(web3, addr)
+    let output = await recompile(data.metadata, data.sources)
+    if (output.metadata == data.metadataRaw) {
+        console.log("Metadata matches!")
+    } else {
+        console.log("Metadata does NOT match!")
+        console.log("From chain:")
+        console.log(data.metadataRaw)
+        console.log("After recompilation:")
+        console.log(output.metadata)
+        process.exit(1)
+    }
+    if (output.deployedBytecode == data.bytecode) {
+        console.log("Deployed bytecode matches! (Note that constructor can still be different!)")
+    } else {
+        console.log("Deployed bytecode does NOT match!")
+        console.log("From chain:")
+        console.log(data.bytecode)
+        console.log("After recompilation:")
+        console.log(output.deployedBytecode)
+        process.exit(1)
+    }
+
+    console.log('')
+
+    if (payload.substr(0, 2) == '0x') {
+        payload = payload.substr(2)
+    }
+    let userdoc = data.metadata['output']['userdoc']['methods']
+    for (var signature in userdoc) {
+        var selector = web3.utils.keccak256(signature).substr(2, 8)
+        if (payload.substr(0, 8) == selector) {
+            console.log("You are calling the function")
+            console.log("  " + signature)
+            console.log("Description: \"" + userdoc[signature]['notice'] + '\"')
+            return
+        }
+    }
+    console.log("Function with selector " + payload.substr(0, 8) + " not found!")
+    console.log("Be careful with this transaction!")
+    process.exit(1)
+}
+
+run(process.argv[2], process.argv[3])
+    .catch(console.log)
```

Comment:
@edisinovcic Just leaving some notes about what "verify.js" V2 could be like since we had further discussion about this in the March 15 & March 23 meetings. 

Think there were two applications:
+ a transaction validator with optional recompilation. This would be a mostly useful as an API method for wallets. It translates raw tx data into human readable output via ABI and userdocs. @ligi has done some work on this already and that could be developed into a lib method published here (via `dist/` 

+ a cli tool that recompiles (and saves?) sources using metadata as an input. It's purpose is to provide on non-centralized (local) version of the deployed service (without running a full monitor).  

Most methods in the in old verify.js file have been extracted to `src/utils.ts`. 

+ reformatMetadata
+ recompile
+ getBytecode
+ cborDecode

Old `verify.js` is missing support for ipfs.  There's relevant code in the Monitor at 
+ [retrieveCode][2]
+ [retrieveMetadataByStorageProvider][3]

[2]: https://github.com/ethereum/source-verify/blob/ede2477cf14c090f862011e2697fc668c448fe0b/src/monitor.ts#L245-L293
[3]: https://github.com/ethereum/source-verify/blob/ede2477cf14c090f862011e2697fc668c448fe0b/src/monitor.ts#L333-L379
Cli interface might be like?
```
metadata-verifier --metadata <path> [--outputDir <path>]
metadata-verifier --address <address> --chain <chain>  [--outputDir <path>]
metadata-verifier --ipfsHash <ipfs-hash> [--ipfsProvider <url>]
```
 
We could use [yargs][1] to help with writing this...

[1]: https://www.npmjs.com/package/yargs